core: migrate stdcm to new infra

eckter opened the pull request:
Change summary:
* Migrate train simulation code to work with the new infra interfaces
* Migrate STDCM to work with the new infra interfaces
* Run the STDCM pathfinding on blocks instead of routes
* Migrate tests, using a new class to build simplified infras
* Cleanup some legacy code (not all of it)

Note: the STDCM changes are fairly minimal (mostly renames and simple replacements). Most of the changes are STDCM dependencies, especially standalone sim. This PR can be reviewed by someone who isn't familiar with the STDCM module. 

Parts of it have already been reviewed in https://github.com/osrd-project/osrd/pull/5017, but it sounded more realistic to squash everything before the rebase.

Erashin committed:
core: migrate stdcm to new infra

Change summary:
* Migrate train simulation code to work with the new infra interfaces
* Migrate STDCM to work with the new infra interfaces
* Run the STDCM pathfinding on blocks instead of routes
* Migrate tests, using a new class to build simplified infras
* Cleanup some legacy code (not all of it)

Khoyo commented on core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMapImpl.kt:
```diff
@@ -172,4 +199,4 @@ data class DistanceRangeMapImpl<T>(
             return res
         }
     }
-}
\ No newline at end of file
+}
```
Me like :)

multun commented on core/kt-osrd-sim-infra/build.gradle:
```diff
@@ -18,6 +18,7 @@ dependencies {
     // PLEASE ADD AND UPDATE DEPENDENCIES USING libs.versions.toml
     implementation project(':kt-fast-collections')
     implementation project(path: ':osrd-geom')
+    implementation project(path: ':envelope-sim')
```
I'd rather not have this dependency. It's only used in a single place for a java adapter that could be moved somewhere else

multun commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/RawSignalingInfra.kt:
```diff
@@ -44,4 +44,14 @@ fun RawSignalingInfra.getLogicalSignalName(signal: LogicalSignalId): String? {
 
+/** Returns the length of a route  */
+@JvmName("getRouteLength")
+fun RawSignalingInfra.getRouteLength(route: RouteId): Distance {
```
I'd rather precompute this, just like we do for blocks

Erashin replied:
Commit to check, route length computed during `resolveReferences` pre `RawInfraBuilderImpl.build()` method.

multun commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/BlockInfraImpl.kt:
```diff
@@ -121,3 +122,21 @@ class BlockInfraImpl(
 }
+
+@JvmName("getBlockEntry")
+fun BlockInfra.getRouteEntry(rawInfra: RawInfra, block: BlockId): DirDetectorId {
```
```suggestion
fun BlockInfra.getBlockEntry(rawInfra: RawInfra, block: BlockId): DirDetectorId {
```

multun commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/BlockInfraImpl.kt:
```diff
@@ -121,3 +122,21 @@ class BlockInfraImpl(
+}
+
+@JvmName("getBlockExit")
+fun BlockInfra.getRouteExit(rawInfra: RawInfra, block: BlockId): DirDetectorId {
```
```suggestion
fun BlockInfra.getBlockExit(rawInfra: RawInfra, block: BlockId): DirDetectorId {
```

axrolld reviewed (approved):
Physics part LGTM 🚀

multun commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraImpl.kt:
```diff
@@ -227,10 +230,35 @@ class RawInfraImpl(
         return res
     }
 
+    override fun getTrackChunkElectricalProfile(
```
I don't think it should be part of this API as is:
 1) either electrical profiles are part of the sim-infra API, and there is no mapping argument
 2) or it's something entirely separate, which thus does not belong in the API implementation

I'd rather have option 2, as this implementation can easily be moved outside of sim-infra

eckter replied:
The purpose was to use the projection tools in PathProperties, but I agree that it's awkward here. I'll make the projection tools public and move this to the place it's called

multun commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/utils/EnvelopeTrainPathUtils.kt:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Builds the ElectrificationMap
+ */
+fun buildElectrificationMap(path: PathProperties): DistanceRangeMap<Electrification> {
```
this does not belong here. it creates a dependency from sim-infra to envelope-sim, which I would like to avoid at all costs. Please move it in another module, such as the main code base.

multun commented on core/src/main/java/fr/sncf/osrd/api/ElectricalProfileSetManager.java:
```diff
@@ -1,7 +1,7 @@
 
 import com.squareup.moshi.JsonDataException;
-import fr.sncf.osrd.external_generated_inputs.ElectricalProfileMapping;
+import fr.sncf.osrd.kt_external_generated_inputs.ElectricalProfileMapping;
```
what's the point of renaming the module? is it not the same?

Erashin replied:
No idea, I don't see a lot of Kotlin package names starting with `kt_` either. @eckter any idea ?

multun commented on core/src/main/java/fr/sncf/osrd/api/pathfinding/RemainingDistanceEstimator.java:
```diff
@@ -77,15 +78,15 @@ public static double minDistanceBetweenSteps(
-            res = Double.min(res, point.distanceAsMeters(target));
-        return res + remainingDistance;
+            resMeters = Double.min(resMeters, point.distanceAsMeters(target));
+        return Distance.fromMeters(resMeters) + remainingDistance;
```
what's going on here? is the double return value used to store a distance in millimeters intentionally? That rings a bell but I'm not 100% sure

multun commented on core/src/main/java/fr/sncf/osrd/api/pathfinding/constraints/ElectrificationConstraints.java:
```diff
@@ -51,7 +51,8 @@ private static Set<Pathfinding.Range> getBlockedRanges(RollingStock stock, PathP
 
                 for (var blockingRange : blockingRanges)
-                    res.add(new Pathfinding.Range(blockingRange.lowerEndpoint(), blockingRange.upperEndpoint()));
+                    if (blockingRange.lowerEndpoint() < blockingRange.upperEndpoint())
```
I'm not sure I understand this one: how can that happen? I wish there were a comment explaining it

eckter replied:
0-length ranges can happen because of the way we build the `RangeSet`, I believe this was already the case before but it wasn't caught. 

But I've fixed it by using closed intervals instead when building the set, it makes more sense. I'll change that `if` into an `assert`.

multun commented on core/src/main/java/fr/sncf/osrd/conflicts/Conflicts.kt:
```diff
@@ -4,8 +4,8 @@ import com.carrotsearch.hppc.IntArrayList
 import com.squareup.moshi.Json
 import fr.sncf.osrd.api.ConflictDetectionEndpoint.ConflictDetectionResult.Conflict
 import fr.sncf.osrd.api.ConflictDetectionEndpoint.ConflictDetectionResult.Conflict.ConflictType
-import fr.sncf.osrd.standalone_sim.result.ResultTrain.SpacingRequirement
```
there aren't any functional changes in this file, I'd rather not touch it if you don't mind

Erashin replied:
Ngl, i disagree. Otherwise we'll never clean up the imports. This should have been done in the PR modifying this file, absolutely, but it wasn't, and I'd rather correct it now than never.

multun commented on core/src/main/java/fr/sncf/osrd/envelope_sim_infra/MRSP.java:
```diff
@@ -1,70 +1,56 @@
+        var pathLength = toMeters(path.getLength());
 
-        // add a limit for the maximum speed the hardware is rated for
+        // Add a limit for the maximum speed the hardware is capable of
```
that's not technically correct, the max speed in the rolling stock definition is not what the train is capable of, it's lower. It's the maximum operational speed.

multun commented on core/src/main/java/fr/sncf/osrd/envelope_sim_infra/MRSP.java:
```diff
@@ -1,70 +1,56 @@
-                var speed = speedRange.getValue().getSpeedLimit(tag);
-                if (Double.isInfinite(speed) || speed == 0)
-                    continue;
+        var offset = addRollingStockLength ? rollingStock.length : 0.;
+        var speedLimits = path.getSpeedLimits(trainTag);
+        for (var speedLimit: speedLimits) {
+            // Compute where this limit is active from and to
+            var start = toMeters(speedLimit.getLower());
+            var end = Math.min(pathLength, offset + toMeters(speedLimit.getUpper()));
+            var speed = toMetersPerSecond(rangeMapEntryToSpeed(speedLimit));
+            if (speed != 0)
```
have you checked it does not change simulation results? not just here, in general

eckter replied:
We didn't run a fuzzer + before/after comparaison, but these parts have lots of good unit tests. We also run the fuzzer looking for 5xx errors.

multun commented on core/src/main/java/fr/sncf/osrd/standalone_sim/SignalProjection.kt:
```diff
@@ -2,54 +2,54 @@ package fr.sncf.osrd.standalone_sim
-        routes.add(route)
-    }
+    // get a new generation route path
+    val routePath = toRouteIdList(PathfindingResultConverter.chunksToRoutes(rawInfra, KtToJavaConverter.toIntList(trainPath.chunks)))
```
I'm surprised we have to guess the route path, didn't we have it?

Erashin replied:
I don't really understand the question. At this point, the only thing we have is the original `RJSTrainPath` (which contains the route path, but in terms of RJSRoute, which is not what we want here) and the `PathProperties`, which does not contain the route path, but contains the path's track chunks (from which we compute the route path here). Do you see another simpler way to do it ?

multun replied:
As discussed, we need to do something like `val routePath = rjsRoutePath.map { infra.getRouteByName(it.id) }`

multun commented on core/src/main/java/fr/sncf/osrd/stdcm/graph/AllowanceManager.java:
```diff
@@ -58,32 +60,32 @@ private double getNeededDelay(STDCMEdge oldEdge) {
             previousEnd = end;
         }
-        assert Math.abs(previousEnd - totalEnvelope.getEndPos()) < 1e-5;
+        assert Math.abs(toMeters(previousEnd) - totalEnvelope.getEndPos()) < 1e-5;
```
are we sure the 1e-5 margin still holds?

Erashin replied:
Indeed, this should be replaced by 1e-2.

multun commented on core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraAdapter.kt:
```diff
@@ -271,36 +282,48 @@ private fun makeChunk(
-                Distance.fromMeters(entry.key.upperEndpoint()),
-                SimNeutralSection(legacyNeutralSection.lowerPantograph)
-            )
+        for ((rangeMap, isAnnouncement) in listOf(
+            Pair(range.neutralSections, false),
+            Pair(range.neutralSectionAnnouncements, true),
+        )) {
+            for (entry in rangeMap.asMapOfRanges()) {
+                val legacyNeutralSection = entry.value
+                res.put(
+                    Distance.fromMeters(entry.key.lowerEndpoint()),
+                    Distance.fromMeters(entry.key.upperEndpoint()),
+                    SimNeutralSection(legacyNeutralSection.lowerPantograph, isAnnouncement)
+                )
+            }
```
this stuff seems unrelated

Erashin replied:
We're now taking into account both the actual neutral sections as well as the announced neutral sections. The change makes sense to me. What do you mean exactly ?

multun commented on core/src/main/kotlin/fr/sncf/osrd/utils/JavaInteroperabilityTools.kt:
```diff
@@ -0,0 +1,20 @@
+/** Java can't handle generics with value classes, it's impossible to extract the underlying speed from a
+ * `RangeMapEntry` in pure java. This method handles this specific case.  */
+fun rangeMapEntryToSpeed(entry: DistanceRangeMap.RangeMapEntry<Speed>): Long {
+    return entry.value.millimetersPerSecond.toLong()
```
isn't entry.value a long in java?

Erashin replied:
In Java, if you have a `DistanceRangeMap<Speed>`, then the value of `DistanceRangeMap.RangeMapEntry<Speed> entry` is not converted to a long, it stays as Speed. Which is why it needs to be converted (and `toMetersPerSecond()` does not work, as Java expects a long as an argument, but would here get a Speed).

eckter committed:
core: electrification: fix range set initialization

multun commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/PathProperties.kt:
```diff
@@ -36,11 +40,27 @@ interface PathProperties {
+    val res = getTrackLocationOffset(location)
+    if (res != null)
+        return res
+    throw OSRDError(ErrorType.InvalidScheduleTrackLocationNotIncludedInPath)
```
The error type if specific to the call site of this function: not all callers may want to yield an InvalidScheduleTrackLocationNotIncludedInPath error. I'd rather move this helper in a context closer to its use, rather than have it in PathProperties.kt

eckter replied:
The function as a whole is a java/kt artifact for one specific call site. It can't be located in a .java file (a nullable inline value type isn't something we can handle).

The alternative would be to throw a generic error which would then be instantly caught and rethrown as an OSRDError. I'm not quite convinced it's better but I don't have a strong opinion on this, what do you think?

multun replied:
I found a way to make it work without a stub:

```diff
diff --git a/core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/PathProperties.kt b/core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/PathProperties.kt
index f348a94a3..bbf9b44e9 100644
--- a/core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/PathProperties.kt
+++ b/core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/PathProperties.kt
@@ -46,6 +46,7 @@ interface PathProperties {
     fun getLength(): Distance
     @JvmName("getTrackLocationAtOffset")
     fun getTrackLocationAtOffset(pathOffset: Distance): TrackLocation
+    @JvmName("getTrackLocationOffset")
     fun getTrackLocationOffset(location: TrackLocation): Distance?
     fun <T> getRangeMapFromUndirected(getData: (chunkId: TrackChunkId) -> DistanceRangeMap<T>): DistanceRangeMap<T>
 
@@ -54,15 +55,6 @@ interface PathProperties {
     val beginOffset: Distance
 }
 
-/** Wraps the method without returning an optional, which can't be handled in java for value types */
-@JvmName("getTrackLocationOffsetOrThrow")
-fun PathProperties.getTrackLocationOffsetOrThrow(location: TrackLocation): Distance {
-    val res = getTrackLocationOffset(location)
-    if (res != null)
-        return res
-    throw OSRDError(ErrorType.InvalidScheduleTrackLocationNotIncludedInPath)
-}
-
 /** Build a Path from chunks and offsets, filtering the chunks outside the offsets */
 @JvmName("buildPathPropertiesFrom")
 fun buildPathPropertiesFrom(
diff --git a/core/src/main/java/fr/sncf/osrd/railjson/parser/RJSStopsParser.java b/core/src/main/java/fr/sncf/osrd/railjson/parser/RJSStopsParser.java
index 08764692a..54f26e982 100644
--- a/core/src/main/java/fr/sncf/osrd/railjson/parser/RJSStopsParser.java
+++ b/core/src/main/java/fr/sncf/osrd/railjson/parser/RJSStopsParser.java
@@ -1,7 +1,5 @@
 package fr.sncf.osrd.railjson.parser;
 
-import static fr.sncf.osrd.sim_infra.api.PathPropertiesKt.getTrackLocationOffsetOrThrow;
-
 import fr.sncf.osrd.railjson.schema.schedule.RJSTrainStop;
 import fr.sncf.osrd.reporting.exceptions.ErrorType;
 import fr.sncf.osrd.reporting.exceptions.OSRDError;
@@ -28,9 +26,12 @@ public class RJSStopsParser {
             long positionMM;
             if (stop.position != null)
                 positionMM = Distance.fromMeters(stop.position);
-            else
-                positionMM = getTrackLocationOffsetOrThrow(path,
-                        RJSTrackLocationParser.parse(infra, stop.location));
+            else {
+                var offset = path.getTrackLocationOffset(RJSTrackLocationParser.parse(infra, stop.location));
+                if (offset == null)
+                    throw new OSRDError(ErrorType.InvalidScheduleTrackLocationNotIncludedInPath);
+                positionMM = offset.getMillimeters();
+            }
 
             if (positionMM > path.getLength())
                 throw new OSRDError(ErrorType.InvalidScheduleOutsideTrainStopPosition);
```

multun commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraBuilder.kt:
```diff
@@ -3,11 +3,61 @@ package fr.sncf.osrd.sim_infra.impl
 import fr.sncf.osrd.fast_collections.mutableIntArrayListOf
 import fr.sncf.osrd.geom.LineString
 import fr.sncf.osrd.reporting.exceptions.OSRDError
-import fr.sncf.osrd.sim_infra.api.*
```
is there a reason why this was removed?

multun commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraBuilder.kt:
```diff
@@ -3,11 +3,61 @@ package fr.sncf.osrd.sim_infra.impl
+import fr.sncf.osrd.sim_infra.api.increasing
 import fr.sncf.osrd.utils.DirectionalMap
 import fr.sncf.osrd.utils.DistanceRangeMap
-import fr.sncf.osrd.utils.indexing.*
-import fr.sncf.osrd.utils.units.*
```
same

multun commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraImpl.kt:
```diff
@@ -3,10 +3,58 @@ package fr.sncf.osrd.sim_infra.impl
 import fr.sncf.osrd.geom.LineString
 import fr.sncf.osrd.reporting.exceptions.ErrorType
 import fr.sncf.osrd.reporting.exceptions.OSRDError
-import fr.sncf.osrd.sim_infra.api.*
-import fr.sncf.osrd.utils.*
-import fr.sncf.osrd.utils.indexing.*
-import fr.sncf.osrd.utils.units.*
```
same

eckter committed:
tests: fuzzer: fix train schedule request format

eckter committed:
core: ScheduleMetadata: handle case where first block isn't used

eckter committed:
core: fix rjs path parsing when going backwards

eckter committed:
core: remove assertion in BlockRecovery

There is already a non-empty assert in all call sites.
For STDCM preprocessing this error is triggered too often
(because of poor signaling data in imported infras),
and an empty result isn't a problem in this case
